File: Codes/p11/p11(C++)/p11.cpp
#include<iostream.h>
#include<math.h>
#include <string>
double temp[4][5];
double temp1[5][6];
using  std::string;

double compEntropy(double a[][5])           //熵
{
  double sum = 0;
  int n = sizeof(a) / sizeof(double);
  for(int i = 0;i < 4;i++)
    for(int j = 0;j < 5;j++)
    sum+=-a[i][j]*(log(a[i][j])/log(2));
  return sum;
}

void output(double a[][5],int nRow,int nCol)     //矩阵输出
{
  for(int i = 0;i < nRow;i++)
  {
    for(int j = 0;j < nCol;j++)
      cout<<a[i][j]<<"\t";
    cout<<endl;
  }
}


double max(double q[][5],int &m,int &n)      //打印最大值和下标
{
	double maxv = 0;
	for(int i = 0;i < 4;i++)
      for(int j = 0;j < 5;j++)
	  {
	 	  if(maxv < q[i][j])
		  { 
			maxv = q[i][j];
			m = i;
			n = j;
		  }

	  }
	return maxv;
}

void move(double q[][5],int nRow,int nCol,int u[][2],int k)     //move函数
{
  int m,n;
  for(int i = 1;i < (nRow + 1);i++)
    for(int j = 1;j < (nCol + 1);j++)
	{	
		if((i-1-u[k][0]) < 0)                        
			m = (i-1-u[k][0]) + 8;
		else
			m = abs(i-1-u[k][0]);

		if((j-1-u[k][1]) < 0)
			n = (j-1-u[k][1]) + 10;
		else
			n = abs(j-1-u[k][1]);
       temp1[i][j] = 0.8*q[m % 4 ][n % 5 ]+0.2* q[i-1][j-1];  //取模
		
	}
  for(int ii = 0;ii < 4;ii++)
    for(int j = 0 ;j < 5;j++)
		temp[ii][j] = temp1[ii+1][j+1];

		
}

void sense(double q[][5],string z,string world1[][5])     //sense函数
{
  int hit;
  for(int i = 0;i < 4;i++)
    for(int j = 0;j < 5;j++)
    {
		if(z == world1[i][j])
			hit = 1;
		else 
			hit = 0;
	
	temp[i][j] = q[i][j] *(hit * 0.7 + (1-hit) * (1 - 0.7)); 

    }
}



void  main()
{
	string world[4][5] ={{"red", "green", "green", "red" , "red"},
       					 {"red", "red", "green", "red",  "red"},
       					 {"red", "red", "green", "green",  "red"},
        				 {"red", "red", "red", "red",  "red"}};
	int nRow = sizeof(world) / sizeof(world[0]);  
	int nCol = sizeof(world[0]) / sizeof(world[0][0]);
	int motions[][2] = {{0, 0}, {0, 1}, {1, 0}, {1, 0},{0, 1}};
	int eCol = sizeof(motions) / sizeof(motions[0]);
	double entropy[2][5];
	double sum1 = 0;
    string measurements[] = {"green", "green", "green" ,"green", "green"};
    double p[4][5],p0[4][5];
   	
	int r_index = 0,c_index = 0;
	double maxV;
	for(int i = 0;i < nRow;i++)
		for(int j = 0;j<nCol;j++)
		p[i][j] = (double)1/(nRow*nCol);

	for(int ii=0;ii<2;ii++)
	  for(int j=0;j<eCol;j++)
		entropy[ii][j] = 0;

	for(int k=0;k<eCol;k++)
	  {
		move(p,nRow,nCol,motions,k);     	//计算move
		for(int i=0;i<nRow;i++)
		  for(int j=0;j<nCol;j++)
			{
			  p[i][j] = temp[i][j];
			  p0[i][j] = p[i][j];
			}

	sense(p,measurements[i], world);   //计算sense
	for(int ii=0;ii<nRow;ii++)         
            for(int j=0;j<nCol;j++)
		{
		p[ii][j] = temp[ii][j];
		sum1+= p[ii][j];
		}
	
	for(int i1=0;i1<nRow;i1++)
            for(int j=0;j<nCol;j++)
				p[i1][j] = (double)p[i1][j]/sum1; 
	sum1 = 0;
       
 //   entropy[0][i]=compEntropy(p0);
 //   entropy[1][i]=compEntropy(p);


	 cout<<"step"<<k<<"The probability before sensing:"<<endl;
	 output(p0,nRow,nCol);
	 cout<<"step"<<k<<"The probability after sensing:"<<endl;
	 output(p,nRow,nCol);
	}
	maxV=max(p,r_index,c_index);
	cout<<"The largest probability "<<maxV<<" occurs at cell"<<"["<<r_index<<"]["<<c_index<<"]"<<endl;


}

File: Codes.python/P11/P11.py
#-------by HYH -------#
import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D
##
world=np.array([['red','green','green','red',  'red'],
                ['red','red',  'green','red',  'red'],
                ['red','red',  'green','green','red'],
                ['red','red',  'red',  'red',  'red']])
color=np.array([['r','b','g','w','y'],
	            ['r','b','g','w','y'],
	            ['r','b','g','w','y'],
	            ['r','b','g','w','y']])
nRow,nCol=np.shape(world)
stop=[0,0]
right=[0,1]
left=[0,-1]
down=[1,0]
up=[-1,0]
pMoveCorrect=0.8
pSenseCorrect=0.7
compEntropy=lambda p:-np.sum(np.sum(p*np.log2(p)))
p=1/(nRow*nCol)*np.ones([nRow,nCol])
##
motions=[stop, right, down, down, right]
measurements=['green','green','green','green','green']
##
def sense(p,z,world,pSenseCorrect):
	nRow,nCol=np.shape(p)
	q=np.zeros([nRow,nCol])
	for r in range(nRow):
		for c in range(nCol):
			if z==world[r][c]:
				hit=1
			elif z!=world[r][c]:
				hit=0
			q[r][c]=p[r][c]*(hit*pSenseCorrect+(1-hit)*(1-pSenseCorrect))
	q=q/sum(sum(q))
	return q

##
def move(p,u,pMoveCorrect):
	nRow,nCol=np.shape(p)
	q=np.zeros([nRow,nCol])
	for r in range(nRow):
		for c in range(nCol):
			q[r][c]=pMoveCorrect*p[(r-u[0])%nRow][(c-u[1]%nCol)]+(1-pMoveCorrect)*p[r][c]
	return q
##
def maxMat(g):
	r=g.argmax(axis=0)
	temp=g[r,range(g.shape[1])]
	index_Row=np.argsort(temp)
	c=index_Row[len(index_Row)-1]
	r=r[c]
	return r,c
##
assert len(motions)==len(measurements),'The variable ''motions'' should be of the same size as ''measurements''!'
entropy=np.zeros([2,len(motions)])	
##
plt.ion()
fig1=plt.figure(figsize=(10,10),dpi=80)
_x=np.arange(1,5)
_y=np.arange(1,6)
_xx,_yy=np.meshgrid(_x,_y)
x,y=_xx.ravel(),_yy.ravel()
for i in range(len(motions)):
	ax=fig1.add_subplot(2,1,1, projection='3d')
	ax.cla()
	p=move(p,motions[i],pMoveCorrect)
	p0=p
	p=sense(p,measurements[i],world,pSenseCorrect)
	p1=p
	entropy[:,i]=[compEntropy(p0),compEntropy(p1)]
	p0=p0.T 
	p1=p1.T
	color0=color.T
	p0=p0.reshape(1,-1)
	p0=p0.tolist()
	p0=sum(p0,[])
	p1=p1.reshape(1,-1)
	p1=p1.tolist()
	p1=sum(p1,[])
	color0=color0.reshape(1,-1)
	color0=color0.tolist()
	color0=sum(color0,[])
	ax.bar3d(x,y,z=np.zeros_like(p0),dx=1,dy=1,dz=p0,color=color0,shade=True)
	ax.set_title('Step %s\n The probability before sensing'%(i+1))
	ax.set_zlim3d(0,0.3)
	ax.set_ylim3d(0.5,4.5)
	ax=fig1.add_subplot(2,1,2, projection='3d')
	ax.bar3d(x,y,z=np.zeros_like(p1),dx=1,dy=1,dz=p1,color=color0,shade=True)
	plt.title('The probability after sensing')
	ax.set_zlim3d(0,0.3)
	ax.set_ylim3d(0.5,4.5)
	plt.pause(1)
##
print('The Posterior:','\n',p)
r,c=maxMat(p)
print('The largest probability %s occurs at cell (%s, %s)\n'%(p[r][c],r,c))
fig2=plt.figure(figsize=(10,10),dpi=80)
plt.plot(range(1,len(motions)+1),entropy[0,:],'bo',MarkerSize=10,label='After sensing')
plt.plot(range(1,len(motions)+1),entropy[1,:],'rx',MarkerSize=10,label='Before sensing')
plt.xlim(0,len(motions)+1)
plt.xlabel('Step')
plt.ylabel('Entropy')
plt.legend()
plt.ioff()
plt.show()